File: desktop/src/com/inkostilation/pong/desktop/display/LobbyScreen.java
package com.inkostilation.pong.desktop.display;

import com.badlogic.gdx.*;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.viewport.ScreenViewport;
import com.inkostilation.pong.commands.request.RequestConnectionCommand;
import com.inkostilation.pong.commands.request.RequestGameListCommand;
import com.inkostilation.pong.commands.request.RequestNewGameCommand;
import com.inkostilation.pong.desktop.controls.InputSystem;
import com.inkostilation.pong.desktop.network.Network;
import com.inkostilation.pong.engine.PongGame;
import com.inkostilation.pong.exceptions.NoEngineException;
import com.kotcrab.vis.ui.VisUI;
import com.kotcrab.vis.ui.util.adapter.ArrayAdapter;
import com.kotcrab.vis.ui.util.adapter.ArrayListAdapter;
import com.kotcrab.vis.ui.util.adapter.ListAdapter;
import com.kotcrab.vis.ui.util.adapter.SimpleListAdapter;
import com.kotcrab.vis.ui.widget.*;
import com.inkostilation.pong.desktop.display.ObserverTree.LobbyNode;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;

public class LobbyScreen implements Screen {

    private Stage view;
    private InputMultiplexer multiplexer = new InputMultiplexer();
    private ArrayList<String[]> list = new ArrayList<>();

    @Override
    public void show() {
        if (!VisUI.isLoaded()) {
            VisUI.load(VisUI.SkinScale.X1);
        }

        view = new Stage(new ScreenViewport());
        Gdx.input.setInputProcessor(view);


        list.add(new String[]{"test", "second", "123"});
        list.add(new String[]{"1111", "----", "\\./"});
        list.add(new String[]{"test", "second", "123"});

        ArrayListAdapter<String[], HorizontalGroup> test = new ArrayListAdapter<String[], HorizontalGroup>(list) {
            @Override
            protected HorizontalGroup createView(String[] item) {
                HorizontalGroup table = new HorizontalGroup();
                for (String i : item) {
                    VisLabel label = new VisLabel(i);
                    table.addActor(label);
                }
                return table;
            }
        };

        VisTable table = new VisTable();
        table.setFillParent(true);
        view.addActor(table);
        table.row().fill().expandY();

        ObserverTree<VisTable, PongGame> tree = new ObserverTree<VisTable, PongGame>(PongGame.class) {
            @Override
            VisTable toActor(PongGame data) {
                VisTable treeTable = new VisTable();
                treeTable.row().expandX().fillX();
                treeTable.add(new VisLabel(data.getGameState().name)).padRight(300);
                treeTable.add(new VisLabel(data.getPlayersNumber() + ""));
                return treeTable;
            }
        };

        ListView<String[]> view = new ListView<>(test);
        view.setItemClickListener(new ListView.ItemClickListener<String[]>() {
            @Override
            public void clicked(String[] item) {
                System.out.println(Arrays.toString(item));
            }
        });

        VerticalGroup group = new VerticalGroup();
        group.addActor(new VisLabel("test"));
        group.pack();
        view.setHeader(group);

        VisWindow listWindow = new VisWindow("Lobby");
        VisTable listTable = new VisTable();
        listTable.setFillParent(true);
        listTable.row().padTop(20);
        listTable.add(group);
        listTable.row().expandY().expandX().fillX().fill().align(Align.center);
        listTable.add(tree);
        listTable.pack();
        listTable.align(Align.top);

        listWindow.addActor(listTable);
        listWindow.setMovable(false);
        listWindow.getTitleTable().removeActor(listWindow.getTitleLabel());
        listWindow.getTitleTable().row();
        listWindow.getTitleTable().add(new VisLabel("Lobby")).center();
        listWindow.getTitleTable().pack();

        view.rebuildView();
        table.add(listWindow).padTop(10).expandY().fill();

        table.row();
        VisTable bottomTable = new VisTable();
        table.add(bottomTable);

        bottomTable.bottom().padRight(100).padTop(10).padBottom(10);

        VisTextButton refreshButton = new VisTextButton("Refresh");
        refreshButton.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                try {
                    Network.getEngine().sendCommand(new RequestGameListCommand());
                } catch (IOException | NoEngineException e) {
                    e.printStackTrace();
                }
            }
        });

        VisTextButton connectButton = new VisTextButton("Connect");
        connectButton.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                try {
                    if (tree.getSelection().getLastSelected() != null) {
                        int index = tree.getSelection().getLastSelected().getPos();
                        ((Game) Gdx.app.getApplicationListener()).setScreen(new PongScreen());
                        Network.getEngine().sendCommand(new RequestConnectionCommand(index));
                    }
                } catch (IOException | NoEngineException e) {
                    e.printStackTrace();
                }
            }
        });

        VisTextButton newButton = new VisTextButton("New");
        newButton.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                try {
                    ((Game) Gdx.app.getApplicationListener()).setScreen(new PongScreen());
                    Network.getEngine().sendCommand(new RequestNewGameCommand());
                } catch (IOException | NoEngineException e) {
                    e.printStackTrace();
                }
            }
        });

        bottomTable.add(refreshButton).padLeft(50);
        bottomTable.add(connectButton).padLeft(50);
        bottomTable.add(newButton).padLeft(50);
        //table.pack();

        try {
            Network.getEngine().sendCommand(new RequestGameListCommand());
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NoEngineException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0.1f, 0.1f, 0.1f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        view.act();
        view.draw();
    }

    @Override
    public void resize(int width, int height) {

    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {

    }

    @Override
    public void dispose() {

    }
}

File: desktop/src/com/inkostilation/pong/desktop/display/LobbyScreen.kt
package com.inkostilation.pong.desktop.display

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.InputMultiplexer
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.HorizontalGroup
import com.badlogic.gdx.scenes.scene2d.ui.VerticalGroup
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.viewport.ScreenViewport
import com.inkostilation.pong.commands.request.RequestConnectionCommand
import com.inkostilation.pong.commands.request.RequestGameListCommand
import com.inkostilation.pong.commands.request.RequestNewGameCommand
import com.inkostilation.pong.desktop.network.Network
import com.inkostilation.pong.engine.PongGame
import com.inkostilation.pong.exceptions.NoEngineException
import com.kotcrab.vis.ui.VisUI
import com.kotcrab.vis.ui.util.adapter.ArrayListAdapter
import com.kotcrab.vis.ui.widget.*
import com.kotcrab.vis.ui.widget.ListView.ItemClickListener
import java.io.IOException
import java.util.*

class LobbyScreen : Screen {
    private var view: Stage? = null
    private val multiplexer = InputMultiplexer()
    private val list = ArrayList<Array<String>>()
    override fun show() {
        if (!VisUI.isLoaded()) {
            VisUI.load(VisUI.SkinScale.X1)
        }
        view = Stage(ScreenViewport())
        Gdx.input.inputProcessor = view
        list.add(arrayOf("test", "second", "123"))
        list.add(arrayOf("1111", "----", "\\./"))
        list.add(arrayOf("test", "second", "123"))
        val test: ArrayListAdapter<Array<String>, HorizontalGroup> = object : ArrayListAdapter<Array<String?>?, HorizontalGroup?>(list) {
            protected override fun createView(item: Array<String?>): HorizontalGroup {
                val table = HorizontalGroup()
                for (i in item) {
                    val label = VisLabel(i)
                    table.addActor(label)
                }
                return table
            }
        }
        val table = VisTable()
        table.setFillParent(true)
        view!!.addActor(table)
        table.row().fill().expandY()
        val tree: ObserverTree<VisTable?, PongGame?> = object : ObserverTree<VisTable, PongGame>(PongGame::class.java) {
            public override fun toActor(data: PongGame): VisTable {
                val treeTable = VisTable()
                treeTable.row().expandX().fillX()
                treeTable.add(VisLabel(data.getGameState().name)).padRight(300f)
                treeTable.add(VisLabel(data.playersNumber.toString() + ""))
                return treeTable
            }
        }
        val view = ListView(test)
        view.setItemClickListener(ItemClickListener<Array<String?>?> { item -> println(Arrays.toString(item)) })
        val group = VerticalGroup()
        group.addActor(VisLabel("test"))
        group.pack()
        view.header = group
        val listWindow = VisWindow("Lobby")
        val listTable = VisTable()
        listTable.setFillParent(true)
        listTable.row().padTop(20f)
        listTable.add(group)
        listTable.row().expandY().expandX().fillX().fill().align(Align.center)
        listTable.add<ObserverTree<VisTable, PongGame>>(tree)
        listTable.pack()
        listTable.align(Align.top)
        listWindow.addActor(listTable)
        listWindow.isMovable = false
        listWindow.titleTable.removeActor(listWindow.titleLabel)
        listWindow.titleTable.row()
        listWindow.titleTable.add(VisLabel("Lobby")).center()
        listWindow.titleTable.pack()
        view.rebuildView()
        table.add(listWindow).padTop(10f).expandY().fill()
        table.row()
        val bottomTable = VisTable()
        table.add(bottomTable)
        bottomTable.bottom().padRight(100f).padTop(10f).padBottom(10f)
        val refreshButton = VisTextButton("Refresh")
        refreshButton.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                try {
                    Network.getEngine().sendCommand(RequestGameListCommand<Any?>())
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: NoEngineException) {
                    e.printStackTrace()
                }
            }
        })
        val connectButton = VisTextButton("Connect")
        connectButton.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                try {
                    if (tree.selection.lastSelected != null) {
                        val index = tree.selection.lastSelected.pos
                        (Gdx.app.applicationListener as Game).screen = PongScreen()
                        Network.getEngine().sendCommand(RequestConnectionCommand<Any?>(index))
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: NoEngineException) {
                    e.printStackTrace()
                }
            }
        })
        val newButton = VisTextButton("New")
        newButton.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                try {
                    (Gdx.app.applicationListener as Game).screen = PongScreen()
                    Network.getEngine().sendCommand(RequestNewGameCommand<Any?>())
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: NoEngineException) {
                    e.printStackTrace()
                }
            }
        })
        bottomTable.add(refreshButton).padLeft(50f)
        bottomTable.add(connectButton).padLeft(50f)
        bottomTable.add(newButton).padLeft(50f)
        //table.pack();
        try {
            Network.getEngine().sendCommand(RequestGameListCommand<Any?>())
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: NoEngineException) {
            e.printStackTrace()
        }
    }

    override fun render(delta: Float) {
        Gdx.gl.glClearColor(0.1f, 0.1f, 0.1f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
        view!!.act()
        view!!.draw()
    }

    override fun resize(width: Int, height: Int) {}
    override fun pause() {}
    override fun resume() {}
    override fun hide() {}
    override fun dispose() {}
}